res/controllers/Pioneer-DDJ-400-script.js

```diff
@@ -0,0 +1,774 @@
+PioneerDDJ400.bendScale = 0.5;
+
+PioneerDDJ400.pointJumpSpace = 0.005; // amount in percent of the Song we can jump back to previous Cue or loop point
+
+PioneerDDJ400.tempoRanges = [0.06, 0.10, 0.16, 0.25]; // WIDE = 25%?
+
+// Keyboard Mode Variables and Settings
+PioneerDDJ400.keyboardHotCuePoint = [0, 0]; // selected HotCue point (eg. PAD) in Keyboard mode per Deck 0 = unset
+PioneerDDJ400.keyboardModeRefCount = [0, 0]; // count the currently pressed Pads per Deck
+PioneerDDJ400.halftoneToPadMap = [4, 5, 6, 7, 0, 1, 2, 3];
+
+// Beatjump Pads
+PioneerDDJ400.beatjumpPad = [-1, 1, -2, 2, -4, 4, -8, 8]; // < 0 = left; else right
+PioneerDDJ400.beatjumpMultiplier = [1, 1]; // Beatjump Beatcount per Deck
+
+// Hotcue Pads saved Loop points
+PioneerDDJ400.hotcueLoopPoints = {
+    "[Channel1]": [],
+    "[Channel2]": []
+};
+
+// Loop Section
+PioneerDDJ400.loopin4beat = [false, false]; // inn4loop is pressed
+PioneerDDJ400.loopout = [false, false]; // out loop is pressed
+PioneerDDJ400.loopAdjustMultiply = 5;
+
+PioneerDDJ400.samplerCallbacks = [];
+
+// Wrapper to easily ignore the function when the button is released.
+var ignoreRelease = function(fn) {
+    "use strict";
+    return function(channel, control, value, status, group) {
+        if (value === 0) { // This means the button is released.
+            return;
+        }
+        return fn(channel, control, value, status, group);
+    };
+};
+
+
+PioneerDDJ400.init = function() {
+    "use strict";
+    // init controller
+    // init tempo Range to 10% (default = 6%)
+    engine.setValue("[Channel1]", "rateRange", PioneerDDJ400.tempoRanges[1]);
+    engine.setValue("[Channel2]", "rateRange", PioneerDDJ400.tempoRanges[1]);
+
+    // show focus buttons on Effect Ract 1 only
+    engine.setValue("[EffectRack1_EffectUnit1]", "show_focus", 1);
+    engine.setValue("[EffectRack1_EffectUnit2]", "show_focus", 0);
+    engine.setValue("[EffectRack1_EffectUnit3]", "show_focus", 0);
+
+    // Connect the VU-Meter LEDS
+    engine.connectControl("[Channel1]", "VuMeter", "PioneerDDJ400.vuMeterUpdate");
+    engine.connectControl("[Channel2]", "VuMeter", "PioneerDDJ400.vuMeterUpdate");
+
+    // reset vumeter
+    PioneerDDJ400.toggleLight(LightsPioneerDDJ400.deck1.vuMeter, false);
+    PioneerDDJ400.toggleLight(LightsPioneerDDJ400.deck2.vuMeter, false);
+    
+    // DJ3730:  added
+    // enable soft takeover for rate controls
+    engine.softTakeover("[Channel1]", "rate", true);
+    engine.softTakeover("[Channel2]", "rate", true);
+
+
+    // DJ3730:  added
+    // Sampler callbacks
+    for (i = 1; i <= 16; ++i) {
+        PioneerDDJ400.samplerCallbacks.push(engine.makeConnection("[Sampler" + i + "]", "play", PioneerDDJ400.samplerPlayOutputCallbackFunction));
+    }
+
+   // DJ3730:  added
+   // trigger "track loaded" animations when a track is loaded
+   for (i=1; i<=2; i++) {
+        engine.connectControl("[Channel"+i+"]","track_loaded", "PioneerDDJ400.trackLoadedLED"+i);
```

Comment:
Since engine.makeConnection accepts a JS function rather than a string containing the function name, the construction of the function name ("PioneerDDJ400.trackLoadedLED"+i) won't work here, so I'll just refactor the code to omit the loop.  There are only two decks being handled here so the loop is a bit unnecessary, anyway.

Reply:
yeah there's no need for a loop when it's only iterated twice

Reply:
While I was at it, I was able to refactor the code to eliminate some of the 'extra' helper functions that were previously used here, so the implementation is a lot cleaner now also.